File: packages/zod-asyncapi/src/schema-dedup.ts
import type { AnySchema } from "./types.js";

/**
 * Schema deduplication utilities for optimizing generated TypeScript types.
 *
 * This module provides fingerprinting and registry functionality to detect
 * structurally identical schemas and generate them only once, reducing
 * memory usage in consuming TypeScript projects.
 */

/**
 * Recursively sorts an object's keys to create a stable representation.
 * This ensures that {a: 1, b: 2} and {b: 2, a: 1} produce the same fingerprint.
 */
function sortObjectDeep(obj: unknown): unknown {
  if (obj === null || typeof obj !== "object") return obj;
  if (Array.isArray(obj)) return obj.map(sortObjectDeep);

  const sorted: Record<string, unknown> = {};
  const keys = Object.keys(obj as Record<string, unknown>).sort();
  for (const key of keys) {
    sorted[key] = sortObjectDeep((obj as Record<string, unknown>)[key]);
  }
  return sorted;
}

/**
 * Generates a canonical fingerprint for an AsyncAPI schema.
 * Identical schemas will produce identical fingerprints.
 */
export function getSchemaFingerprint(schema: AnySchema): string {
  return JSON.stringify(sortObjectDeep(schema));
}

/**
 * Registry for tracking unique schemas and their canonical names.
 */
export interface SchemaRegistry {
  /** Map from fingerprint to the first schema name that used it */
  fingerprintToName: Map<string, string>;
  /** Map from schema name to its fingerprint (for reverse lookup) */
  nameToFingerprint: Map<string, string>;
}

/**
 * Creates a new empty schema registry.
 */
export function createSchemaRegistry(): SchemaRegistry {
  return {
    fingerprintToName: new Map(),
    nameToFingerprint: new Map(),
  };
}

/**
 * Result of registering a schema.
 */
export interface RegisterSchemaResult {
  /** Whether this is a new unique schema */
  isNew: boolean;
  /** The canonical name for this schema (may be different from input name if duplicate) */
  canonicalName: string;
}

/**
 * Registers a schema in the registry. If an identical schema already exists,
 * returns the existing canonical name instead.
 */
export function registerSchema(
  registry: SchemaRegistry,
  name: string,
  schema: AnySchema,
): RegisterSchemaResult {
  const fingerprint = getSchemaFingerprint(schema);

  const existing = registry.fingerprintToName.get(fingerprint);
  if (existing) {
    return { isNew: false, canonicalName: existing };
  }

  registry.fingerprintToName.set(fingerprint, name);
  registry.nameToFingerprint.set(name, fingerprint);
  return { isNew: true, canonicalName: name };
}

/**
 * Pre-registers a schema with a specific fingerprint.
 * Used for common schemas that should take priority.
 */
export function preRegisterSchema(
  registry: SchemaRegistry,
  name: string,
  fingerprint: string,
): void {
  registry.fingerprintToName.set(fingerprint, name);
  registry.nameToFingerprint.set(name, fingerprint);
}

/**
 * Represents a common schema that appears multiple times.
 */
export interface CommonSchema {
  /** The canonical name for this schema */
  name: string;
  /** The schema definition */
  schema: AnySchema;
  /** The fingerprint for deduplication */
  fingerprint: string;
  /** Number of times this schema appears */
  count: number;
}

/**
 * Scans schemas and identifies those that appear multiple times.
 * Returns common schemas sorted by count (most common first).
 */
export function findCommonSchemas(
  schemas: Array<{ name: string; schema: AnySchema }>,
  minCount: number = 2,
): CommonSchema[] {
  const fingerprints = new Map<
    string,
    { schema: AnySchema; names: string[] }
  >();

  // Count occurrences of each unique schema
  for (const { name, schema } of schemas) {
    const fingerprint = getSchemaFingerprint(schema);
    const existing = fingerprints.get(fingerprint);

    if (existing) {
      existing.names.push(name);
    } else {
      fingerprints.set(fingerprint, {
        schema,
        names: [name],
      });
    }
  }

  // Filter to schemas appearing minCount+ times
  const commonSchemas: CommonSchema[] = [];
  for (const [fingerprint, data] of fingerprints) {
    if (data.names.length >= minCount) {
      // Use first occurrence as the canonical name
      const name = data.names[0]!;

      commonSchemas.push({
        name,
        schema: data.schema,
        fingerprint,
        count: data.names.length,
      });
    }
  }

  // Sort by count descending (most common first)
  return commonSchemas.sort((a, b) => b.count - a.count);
}

File: packages/python-pydantic-openapi/src/alt_stack_pydantic_openapi/schema_dedup.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, TypedDict

from .types import AnySchema


def _sort_object_deep(obj: Any) -> Any:
    if obj is None or not isinstance(obj, (dict, list)):
        return obj
    if isinstance(obj, list):
        return [_sort_object_deep(item) for item in obj]
    sorted_obj: dict[str, Any] = {}
    for key in sorted(obj.keys()):
        sorted_obj[key] = _sort_object_deep(obj[key])
    return sorted_obj


def get_schema_fingerprint(schema: AnySchema) -> str:
    return json.dumps(_sort_object_deep(schema), sort_keys=True)


@dataclass
class SchemaRegistry:
    fingerprint_to_name: dict[str, str]


def create_schema_registry() -> SchemaRegistry:
    return SchemaRegistry(fingerprint_to_name={})


@dataclass
class RegisterSchemaResult:
    is_new: bool
    canonical_name: str


def register_schema(
    registry: SchemaRegistry,
    name: str,
    schema: AnySchema,
) -> RegisterSchemaResult:
    fingerprint = get_schema_fingerprint(schema)
    existing = registry.fingerprint_to_name.get(fingerprint)
    if existing:
        return RegisterSchemaResult(is_new=False, canonical_name=existing)
    registry.fingerprint_to_name[fingerprint] = name
    return RegisterSchemaResult(is_new=True, canonical_name=name)


def pre_register_schema(
    registry: SchemaRegistry,
    name: str,
    fingerprint: str,
) -> None:
    registry.fingerprint_to_name[fingerprint] = name


def extract_error_code(schema: AnySchema) -> str | None:
    props = schema.get("properties")
    if not isinstance(props, dict):
        return None
    error_obj = props.get("error")
    if not isinstance(error_obj, dict):
        return None
    error_props = error_obj.get("properties")
    if not isinstance(error_props, dict):
        return None
    code_schema = error_props.get("code")
    if not isinstance(code_schema, dict):
        return None
    code_enum = code_schema.get("enum")
    if isinstance(code_enum, list) and len(code_enum) == 1:
        return str(code_enum[0])
    return None


def error_code_to_pascal_case(code: str) -> str:
    return "".join(part.capitalize() for part in code.split("_"))


def generate_common_error_schema_name(error_code: str) -> str:
    return f"{error_code_to_pascal_case(error_code)}Error"


@dataclass
class CommonSchema:
    name: str
    schema: AnySchema
    fingerprint: str
    count: int


class NamedSchema(TypedDict):
    name: str
    schema: AnySchema


def find_common_schemas(
    schemas: list[NamedSchema],
    min_count: int = 2,
) -> list[CommonSchema]:
    fingerprints: dict[str, dict[str, Any]] = {}

    for item in schemas:
        name = item["name"]
        schema = item["schema"]
        fingerprint = get_schema_fingerprint(schema)
        existing = fingerprints.get(fingerprint)
        if existing:
            existing["names"].append(name)
        else:
            fingerprints[fingerprint] = {
                "schema": schema,
                "names": [name],
                "error_code": extract_error_code(schema),
            }

    common: list[CommonSchema] = []
    for fingerprint, data in fingerprints.items():
        names = data["names"]
        if len(names) < min_count:
            continue
        error_code = data["error_code"]
        name = (
            generate_common_error_schema_name(error_code)
            if isinstance(error_code, str)
            else names[0]
        )
        common.append(
            CommonSchema(
                name=name,
                schema=data["schema"],
                fingerprint=fingerprint,
                count=len(names),
            )
        )

    return sorted(common, key=lambda item: item.count, reverse=True)
